graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java

```diff
@@ -0,0 +1,356 @@
+            out.close(e);
+        }
+    }
+
+    private void handleBlocking(WebSocketWriter out, String id,
+                                @Nullable ExecutionResult executionResult, @Nullable Throwable t) {
+        if (t != null) {
+            logger.debug("Error handling subscription", t);
+            writeError(out, id, t);
+            return;
+        }
+
+        if (executionResult == null) {
+            logger.debug("ExecutionResult was null but no error was thrown");
+            writeError(out, id, new IllegalArgumentException("ExecutionResult was null"));
+            return;
+        }
+
+        if (!executionResult.getErrors().isEmpty()) {
+            writeError(out, id, executionResult.getErrors());
+            return;
+        }
+
+        if (!(executionResult.getData() instanceof Publisher)) {
+            writeError(out, id, new Exception("Result of operation was not a subscription"));
+            return;
+        }
+
+        final Publisher<ExecutionResult> publisher = executionResult.getData();
+        final StreamMessage<ExecutionResult> streamMessage;
+        if (publisher instanceof StreamMessage) {
+            streamMessage = (StreamMessage<ExecutionResult>) publisher;
+        } else {
+            streamMessage = new PublisherBasedStreamMessage<>(publisher);
+        }
+
+        final ExecutionResultSubscriber executionResultSubscriber =
+                new ExecutionResultSubscriber(id, new GraphqlSubProtocol() {
+                    @Override
+                    public void sendResult(String operationId, ExecutionResult executionResult) {
+                        writeNext(out, operationId, executionResult);
+                    }
+
+                    @Override
+                    public void sendGraphqlErrors(List<GraphQLError> errors) {
+                        writeError(out, id, errors);
+                    }
+
+                    @Override
+                    public void completeWithError(Throwable cause) {
+                        writeError(out, id, cause);
+                        writeComplete(out, id);
+                        final ExecutionResultSubscriber s = graphqlSubscriptions.remove(id);
+                        if (s != null) {
+                            s.setCompleted();
+                            streamMessage.abort(cause);
+                        }
+                    }
+
+                    @Override
+                    public void complete() {
+                        writeComplete(out, id);
+                        final ExecutionResultSubscriber s = graphqlSubscriptions.remove(id);
+                        if (s != null) {
+                            s.setCompleted();
+                            streamMessage.abort();
+                        }
+                    }
+                });
+
+        graphqlSubscriptions.put(id, executionResultSubscriber);
+        streamMessage.subscribe(executionResultSubscriber);
```

Comment:
So that we can guarantee we access the map data structure from a single thread (to prevent concurrent modification exceptions)

```suggestion
        streamMessage.subscribe(executionResultSubscriber, ctx.eventLoop());
```